Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -529,18 +545,10 @@ internal struct RedirectHandler<T> {
         let originalURL = self.request.url
 
         var request = self.request
-        request.url = redirectURL
-
-        if let redirectHost = redirectURL.host {
-            request.host = redirectHost
-        } else {
-            preconditionFailure("redirectURL doesn't contain a host")
-        }
-
-        if let redirectScheme = redirectURL.scheme {
-            request.scheme = redirectScheme
-        } else {
-            preconditionFailure("redirectURL doesn't contain a scheme")
+        do {
+            try request.setUrl(redirectURL)
+        } catch {
+            return promise.fail(error)
```

Comment:
i dont love this, open to other ideas

Reply:
I think this is right.

Reply:
its "right" but not very friendly API imo

Reply:
It's as friendly an API as you can have in Swift when something can fail. There's just really no other good way to structure this code.